### PR TITLE
Feature/whitelists

### PR DESCRIPTION
**Pull request by Alexey-N-Chernyshov**
New BRVS Rules for whitelisting:
1) UpdateWhitlistingRule - updates BRVS record of whitelist and sets validation time.
2) CheckWhitelistRule - validates withdrawal. Checks that address is whitelisted and validation period is expired.

### CODE REVIEW

File: brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/CheckWhitelistRule.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  private static final Logger logger = LoggerFactory.getLogger(UpdateWhitelistRule.class);
+
+  private String withdrawalAccount = "notary@notary";
+
+  private String brvsAccountId;
+  private KeyPair brvsAccountKeyPair;
+  private IrohaAPI irohaAPI;
+
+  public CheckWhitelistRule(String brvsAccountId, KeyPair brvsAccountKeyPair, IrohaAPI irohaAPI) {
+    logger.info("START CheckWhitelistRule");
```
Comment (Mingela) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/CheckWhitelistRule.java]: Do we really need such a message :suspect:

File: brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/CheckWhitelistRule.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        .stream()
+        .filter(Commands.Command::hasTransferAsset)
+        .map(Commands.Command::getTransferAsset)
+        .filter(transfer -> transfer.getDestAccountId().equals(withdrawalAccount))
+        .allMatch(transfer -> {
+              try {
+                String clientId = transfer.getSrcAccountId();
+                String address = transfer.getDescription();
+                String asset = transfer.getAssetId();
+                String assetDomain = asset.substring(asset.lastIndexOf("#") + 1);
```
Comment (Mingela) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/CheckWhitelistRule.java]: it is better to move this function to the Utils class

File: brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/CheckWhitelistRule.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+                String whitelistKey = WhitelistUtils.assetToWhitelistKey.get(assetDomain);
```
Comment (Mingela) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/CheckWhitelistRule.java]: what if it's null?

File: brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/CheckWhitelistRule.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+import org.slf4j.LoggerFactory;
+
+/**
+ * Rule that checks whitelisted addresses on withdrawal.
+ */
+public class CheckWhitelistRule implements Rule {
+
+  private static final Logger logger = LoggerFactory.getLogger(UpdateWhitelistRule.class);
+
+  private String withdrawalAccount = "notary@notary";
```
Comment (Mingela) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/CheckWhitelistRule.java]: can be `private static final`

File: brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/CheckWhitelistRule.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+  private String brvsAccountId;
```
Comment (Mingela) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/CheckWhitelistRule.java]: please make them `final`

File: brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/UpdateWhitelistRule.java [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+/**
+ * A rule that is triggered on a client whitelist update. Checks the difference in new and old
+ * whitelist and updates BRVS records with validation period.
+ */
+public class UpdateWhitelistRule implements Rule {
+
+  private static final Logger logger = LoggerFactory.getLogger(UpdateWhitelistRule.class);
+
+  private String brvsAccountId;
```
Comment (Mingela) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/UpdateWhitelistRule.java]: please make them `final`

File: brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/UpdateWhitelistRule.java [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+  private IrohaAPI irohaAPI;
+
+  /**
+   * When new address will be valid
+   */
+  private long validationPeriod;
+
+  public UpdateWhitelistRule(String brvsAccountId, KeyPair brvsAccountKeyPair, IrohaAPI irohaAPI,
+      long validationPeriod) {
+    logger.info("START UpdateWhitelistRule");
```
Comment (Mingela) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/UpdateWhitelistRule.java]: Do we really need such a message :suspect:

File: brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/CheckWhitelistRule.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+                if (!whitelistValidated.containsKey(address)) {
+                  logger.info("Not allowed. " + address + " not in whitelist.");
+                  return false;
+                }
+
+                if (whitelistValidated.get(address) <= now) {
+                  logger.info("Transfer allowed.");
+                  return true;
+                } else {
+                  logger.info("Not allowed. Address " + address + "will be validated after "
```
Comment (Mingela) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/CheckWhitelistRule.java]: `" will`

File: brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/UpdateWhitelistRule.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+            newWhitelistValidated.putAll(
+                clientWhitelist.stream()
+                    .filter(address -> !oldWhitelistValidated.containsKey(address))
+                    .collect(Collectors.toMap(a -> a, t -> validationTime))
+            );
+
+            if (newWhitelistValidated.equals(oldWhitelistValidated)) {
+              logger.info("No changes in whitelist, nothing to update");
+            } else {
+              String jsonFin = WhitelistUtils.serializeBRVSWhitelist(newWhitelistValidated);
```
Comment (Mingela) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/UpdateWhitelistRule.java]: why is the variable called `Fin`?

File: brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/WhitelistUtils.java [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+   * @param whitelistKey - details key
+   * @return Map of (address to validation time)
+   * @throws Exception on invalid response from Iroha
+   */
+  static Map<String, Long> getBRVSWhitelist(
+      String brvsAccountId,
+      KeyPair brvsAccountKeyPair,
+      IrohaAPI irohaAPI,
+      String clientId,
+      String whitelistKey) throws Exception {
```
Comment (Mingela) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/WhitelistUtils.java]: throws `IllegalAccessException`

File: brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/WhitelistUtils.java [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+  }
+
+  /**
+   * Reverse changes of 'irohaEscape'
+   */
+  public static String irohaUnEscape(String str) {
+    return str.replace(IROHA_FRIENDLY_QUOTE, "\"")
+        .replace(IROHA_FRIENDLY_EOL, "\n");
+  }
+
```
Comment (Mingela) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/WhitelistUtils.java]: empty line 😄

File: brvs-rules/src/test/java/iroha/validation/rules/CheckWhitelistRule.java [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+import iroha.protocol.QryResponses.QueryResponse;
+import iroha.protocol.Queries.Query;
+import iroha.protocol.TransactionOuterClass.Transaction;
+import iroha.validation.rules.impl.whitelist.WhitelistUtils;
+import java.security.KeyPair;
+import java.util.Collections;
+import jp.co.soramitsu.crypto.ed25519.Ed25519Sha3;
+import jp.co.soramitsu.iroha.java.IrohaAPI;
+import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentCaptor;
```
Comment (Mingela) [inline brvs-rules/src/test/java/iroha/validation/rules/CheckWhitelistRule.java]: last 2 imports are not used

File: brvs-rules/src/test/java/iroha/validation/rules/CheckWhitelistRule.java [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+import org.mockito.Captor;
+
+class CheckWhitelistRule {
+
+  private static final Ed25519Sha3 crypto = new Ed25519Sha3();
+
+  private String brvsAccountId = "brvs@brvs";
+  private KeyPair brvsAccountKeyPair = crypto.generateKeypair();
+  private IrohaAPI irohaAPI = mock(IrohaAPI.class);
+  private long validationPeriod = 10;
```
Comment (Mingela) [inline brvs-rules/src/test/java/iroha/validation/rules/CheckWhitelistRule.java]: it is not used

File: brvs-rules/src/test/java/iroha/validation/rules/CheckWhitelistRule.java [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+import jp.co.soramitsu.iroha.java.IrohaAPI;
+import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+
+class CheckWhitelistRule {
+
+  private static final Ed25519Sha3 crypto = new Ed25519Sha3();
+
+  private String brvsAccountId = "brvs@brvs";
```
Comment (Mingela) [inline brvs-rules/src/test/java/iroha/validation/rules/CheckWhitelistRule.java]: make them all `final` please

File: brvs-rules/src/test/java/iroha/validation/rules/CheckWhitelistRule.java [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+  private String withdrawalClientId = "notary@notary";
+
+  private Commands.Command command = mock(Commands.Command.class, RETURNS_DEEP_STUBS);
+  private Transaction transaction = mock(Transaction.class, RETURNS_DEEP_STUBS);
+  private QueryResponse queryResponse = mock(QueryResponse.class, RETURNS_DEEP_STUBS);
+
+  private Rule rule = new iroha.validation.rules.impl.whitelist.CheckWhitelistRule(brvsAccountId,
+      brvsAccountKeyPair, irohaAPI);
+
+  private void setEnvironmentTest(String key, String clientListResponse) {
```
Comment (Mingela) [inline brvs-rules/src/test/java/iroha/validation/rules/CheckWhitelistRule.java]: key is not used

File: brvs-rules/src/test/java/iroha/validation/rules/CheckWhitelistRule.java [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+    String key = WhitelistUtils.ETH_WHITELIST_KEY;
+
+    // BRVS response with whitelist with different address
+    String clientListResponse = "{\"brvs@brvs\" : {\"eth_whitelist\" : \"{\\\"0x6826d84158e516f631bBf14586a9BE7e255b2D22\\\":999999999999}\"}}";
+
+    setEnvironmentTest(key, clientListResponse);
+
+    assertFalse(rule.isSatisfiedBy(transaction));
+  }
+
```
Comment (Mingela) [inline brvs-rules/src/test/java/iroha/validation/rules/CheckWhitelistRule.java]: empty line 😄

File: brvs-rules/src/test/java/iroha/validation/rules/UpdateWhitelistRuleTest.java [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+import jp.co.soramitsu.iroha.java.IrohaAPI;
+import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+
+class UpdateWhitelistRuleTest {
+
+  private static final Ed25519Sha3 crypto = new Ed25519Sha3();
+
+  private String brvsAccountId = "brvs@brvs";
```
Comment (Mingela) [inline brvs-rules/src/test/java/iroha/validation/rules/UpdateWhitelistRuleTest.java]: please make them all `final`

File: brvs-rules/src/test/java/iroha/validation/rules/UpdateWhitelistRuleTest.java [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+    assertEquals(brvsAccountId,
+        captor.getValue().getPayload().getReducedPayload().getCreatorAccountId());
+    assertEquals(1, captor.getValue().getPayload().getReducedPayload().getCommandsList().size());
+    Command cmd = captor.getValue().getPayload().getReducedPayload().getCommandsList().get(0);
+    assertTrue(cmd.hasSetAccountDetail());
+    assertEquals(clientId, cmd.getSetAccountDetail().getAccountId());
+    assertEquals(key, cmd.getSetAccountDetail().getKey());
+
+    Set<String> whitelistSet = new HashSet<>(ethWhitelist);
+    assertTrue(
```
Comment (Mingela) [inline brvs-rules/src/test/java/iroha/validation/rules/UpdateWhitelistRuleTest.java]: `assertEquals`?

File: brvs-rules/src/test/java/iroha/validation/rules/UpdateWhitelistRuleTest.java [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+        WhitelistUtils.deserializeBRVSWhitelist(cmd.getSetAccountDetail().getValue()).keySet()
+            .equals(whitelistSet));
+  }
+
```
Comment (Mingela) [inline brvs-rules/src/test/java/iroha/validation/rules/UpdateWhitelistRuleTest.java]: empty line 😄

File: config/context/application.properties [hunk 20]
```diff
@@ -4,12 +4,13 @@ credential.privkeyFilePath=config/context/priv.key
...
-accounts.holder=notary@notary
+accounts.holder=client_accounts@notary
 iroha.host=brvs-iroha
 iroha.port=50051
 mongo.host=brvs-mongodb
 mongo.port=27017
-rmq.host=brvs-rmq
+rmq.host=d3-rmq
 rmq.port=5672
+whitelist.validation=86400
```
Comment (Mingela) [inline config/context/application.properties]: commit only this change please :)

File: brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/UpdateWhitelistRule.java [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+              logger.info("Send whitelist to Iroha: " + jsonFin);
+
+              irohaAPI.transactionSync(
+                  Transaction.builder(brvsAccountId)
```
Comment (Mingela) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/UpdateWhitelistRule.java]: Please provide `createdTime` equal to the beginning of the day in order to support many BRVS instances in the domain so brvs account could be MST

File: brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/CheckWhitelistRule.java (hunk 3 above)
Comment (Alexey-N-Chernyshov) [reply brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/CheckWhitelistRule.java]: A wrong asset transfer - it is an incorrect state, we will catch Exception.

File: config/context/application.properties (hunk 20 above)
Comment (Alexey-N-Chernyshov) [reply config/context/application.properties]: I suggest to keep 'client_accounts@notary' for accounts.holder..

Comment (Mingela) [reply config/context/application.properties]: Okay 😆

File: brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/UpdateWhitelistRule.java (hunk 21 above)
Comment (Alexey-N-Chernyshov) [reply brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/UpdateWhitelistRule.java]: It is a bad approach to set it equals to the beginning of the day. What if BRVS discovers tx at midnight? Part of multisig transactions will have one date and part will have another. It is better to take time from external source, for example incoming transaction.

Comment (Mingela) [reply brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/UpdateWhitelistRule.java]: Sure you can do it in a such way. My point was just about time sync, not exact approach 👍

Comment (Alexey-N-Chernyshov) [reply brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/UpdateWhitelistRule.java]: done

File: brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/CheckWhitelistRule.java [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+import org.slf4j.LoggerFactory;
+
+/**
+ * Rule that checks whitelisted addresses on withdrawal.
+ */
+public class CheckWhitelistRule implements Rule {
+
+  private static final Logger logger = LoggerFactory.getLogger(UpdateWhitelistRule.class);
+
+  private static final String withdrawalAccount = "notary@notary";
```
Comment (dolgopolovwork) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/CheckWhitelistRule.java]: But I send money to a different account.

Comment (dolgopolovwork) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/CheckWhitelistRule.java]: But I send money to a different account.

File: brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/WhitelistUtils.java [hunk 23]
```diff
@@ -0,0 +1,146 @@
...
+public class WhitelistUtils {
+
+  private static final Logger logger = LoggerFactory.getLogger(WhitelistUtils.class);
+
+
+  public static final String ETH_WHITELIST_KEY = "eth_whitelist";
+  public static final String BTC_WHITELIST_KEY = "btc_whitelist";
+
+  static final Map<String, String> assetToWhitelistKey;
```
Comment (dolgopolovwork) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/WhitelistUtils.java]: can we make it private?

File: brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/UpdateWhitelistRule.java [hunk 24]
```diff
@@ -0,0 +1,130 @@
...
+                      .setCreatedTime(createdTime)
+                      .setAccountDetail(clientId, whitelistKey,
+                          WhitelistUtils.irohaEscape(jsonNewBrvsWhitelist))
+                      .sign(brvsAccountKeyPair)
+                      .build()
+              );
+            }
+          } catch (Exception e) {
+            logger.error("Error during parsing JSON ", e);
```
Comment (dolgopolovwork) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/UpdateWhitelistRule.java]: Confusing message. This is not the only operation that may fail in `try` block, as I can see.

File: brvs-rules/src/test/java/iroha/validation/rules/CheckWhitelistRule.java [hunk 25]
```diff
@@ -0,0 +1,110 @@
...
+import iroha.protocol.QryResponses.QueryResponse;
+import iroha.protocol.Queries.Query;
+import iroha.protocol.TransactionOuterClass.Transaction;
+import java.security.KeyPair;
+import java.util.Collections;
+import jp.co.soramitsu.crypto.ed25519.Ed25519Sha3;
+import jp.co.soramitsu.iroha.java.IrohaAPI;
+import org.junit.jupiter.api.Test;
+
+class CheckWhitelistRule {
```
Comment (dolgopolovwork) [inline brvs-rules/src/test/java/iroha/validation/rules/CheckWhitelistRule.java]: By convention this test must be called `CheckWhitelistRuleTest` and moved to the `iroha.validation.rules.impl.whitelist` package.

File: brvs-rules/src/test/java/iroha/validation/rules/UpdateWhitelistRuleTest.java [hunk 26]
```diff
@@ -0,0 +1,154 @@
+package iroha.validation.rules;
```
Comment (dolgopolovwork) [inline brvs-rules/src/test/java/iroha/validation/rules/UpdateWhitelistRuleTest.java]: The same thing

File: brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/WhitelistUtils.java (hunk 23 above)
Comment (Alexey-N-Chernyshov) [reply brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/WhitelistUtils.java]: Nope, it is used to determine which asset belongs to what whitelist key (sora belongs to ether whitelist).

File: brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/CheckWhitelistRule.java (hunk 22 above)
Comment (Alexey-N-Chernyshov) [reply brvs-rules/src/main/java/iroha/validation/rules/impl/whitelist/CheckWhitelistRule.java]: Moved to context

File: config/context/spring-context.xml [hunk 27]
```diff
@@ -126,9 +126,29 @@
...
+  <bean id="updateWhitelistRule" class="iroha.validation.rules.impl.whitelist.UpdateWhitelistRule">
+    <constructor-arg name="brvsAccountId" value="${credential.accountId}"/>
+    <constructor-arg name="brvsAccountKeyPair" ref="brvsAccountKeyPair"/>
+    <constructor-arg name="irohaAPI" ref="irohaAPI"/>
+    <constructor-arg name="validationPeriod" value="${whitelist.validation}"/>
+  </bean>
+  <bean id="checkEthWhitelistRule" class="iroha.validation.rules.impl.whitelist.CheckWhitelistRule">
+    <constructor-arg name="brvsAccountId" value="${credential.accountId}"/>
+    <constructor-arg name="brvsAccountKeyPair" ref="brvsAccountKeyPair"/>
+    <constructor-arg name="irohaAPI" ref="irohaAPI"/>
```
Comment (dolgopolovwork) [inline config/context/spring-context.xml]: Don't you forget to add `withdrawalAccount`?

Comment (Alexey-N-Chernyshov) [reply config/context/spring-context.xml]: yep, fixed

**Pull request merged**